query/src/sql/planner/binder/select.rs

```diff
@@ -12,306 +12,146 @@
-    }
+#[derive(Debug)]
+pub struct SelectItem<'a> {
+    pub select_target: &'a SelectTarget<'a>,
+    pub scalar: Scalar,
+    pub alias: String,
+}
 
+impl<'a> Binder {
     pub(super) async fn bind_select_stmt(
         &mut self,
         bind_context: &BindContext,
         stmt: &SelectStmt<'a>,
         order_by: &[OrderByExpr<'a>],
     ) -> Result<(SExpr, BindContext)> {
-        let (mut s_expr, from_context) = if let Some(from) = &stmt.from {
+        let (mut s_expr, mut from_context) = if let Some(from) = &stmt.from {
             self.bind_table_reference(bind_context, from).await?
         } else {
             self.bind_one_table(stmt).await?
         };
 
         if let Some(expr) = &stmt.selection {
-            s_expr = self.bind_where(&from_context, expr, s_expr, false).await?;
+            s_expr = self.bind_where(&from_context, expr, s_expr).await?;
         }
 
-        // Output of current `SELECT` statement.
-        let mut output_context = self
+        // Generate a analyzed select list with from context
+        let mut select_list = self
             .normalize_select_list(&from_context, &stmt.select_list)
             .await?;
 
-        let agg_info = self.analyze_aggregate(&output_context)?;
-        if !agg_info.aggregate_functions.is_empty() || !stmt.group_by.is_empty() {
-            (s_expr, output_context) = self
-                .bind_group_by(
-                    &from_context,
-                    output_context,
-                    s_expr,
-                    &stmt.group_by,
-                    &agg_info,
-                )
+        let mut projections = self.analyze_projection(&select_list)?;
+
+        self.analyze_aggregate_select(&mut from_context, &mut select_list)?;
+        let having = if let Some(having) = &stmt.having {
+            Some(
+                self.analyze_aggregate_having(&mut from_context, having)
+                    .await?,
+            )
+        } else {
+            None
+        };
+        let order_items = self.analyze_order_items(&from_context, &select_list, order_by)?;
+
+        if !from_context.aggregate_info.aggregate_functions.is_empty()
+            || !stmt.group_by.is_empty()
+            || stmt.having.is_some()
+        {
+            s_expr = self
+                .bind_aggregate(&mut from_context, &select_list, &stmt.group_by, s_expr)
                 .await?;
-        }
 
-        if let Some(expr) = &stmt.having {
-            s_expr = self.bind_where(&from_context, expr, s_expr, true).await?;
+            if let Some(having) = having {
+                s_expr = self.bind_having(&from_context, having, s_expr).await?;
+            }
         }
 
-        s_expr = self.bind_projection(&output_context, s_expr)?;
-
         if !order_by.is_empty() {
             s_expr = self
-                .bind_order_by(&from_context, &output_context, s_expr, order_by)
+                .bind_order_by(&from_context, order_items, &mut projections, s_expr)
```

Comment:
Why put `order_by` before `projection`?

Reply:
For the query `SELECT a+1 AS b FROM t ORDER BY a, b`, if we build a project `(a+1 -> b)` on the top of query, then there is no `a` in current context.

Therefore if we allow it to choose a invisible column as order item, we have to use a project to remove the invisible column in the end, and use another proejct to generate a `(a+1 -> b)` before `ORDER BY`.

The `bind_order_by` may build a project before sort, to satisfy the requirement.

A logical plan tree of the above query would look like:
```
Project(b)
|
Sort(a, b)
|
Project(a, a+1 -> b)
|
Scan
```

Reply:
make sense